File: BOJ/2473_세+용액.cpp
/**
 * 20-10-01
 * BOJ 2473 세 용액
 * 이분 탐색.. + 자료형(long long)
 * */

#include <iostream>
#include <algorithm>
#define MAX 3000000001
using namespace std;

int N;
long long arr[5001];
long long ans[3];

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0); cout.tie(0);

    cin >> N;
    for(int i=0; i<N; i++)
        cin >> arr[i];

    sort(arr, arr+N);

    long long res = MAX;

    for(int idx=0; idx<N-2; idx++){
        int left = idx+1;
        int right = N-1;

        while(left<right){
            long long sum = arr[idx] + arr[left] + arr[right];

            if(res > abs(sum)){
                res = abs(sum);
                ans[0] = arr[idx];
                ans[1] = arr[left];
                ans[2] = arr[right];
            }

            if(sum==0) break;
            else if(sum<0) left++;
            else right--;
        }
    }
    
    cout << ans[0] << ' ' << ans[1] << ' ' << ans[2];
    return 0;
}

File: BOJ/2473_세+용액.java
/**
 * 22-02-05
 * BOJ 2473 세 용액
 * 투포인터, 정렬
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class test2 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        long[] arr = new long[N];
        long[] ans = new long[3];
        long mn = Long.MAX_VALUE;

        st = new StringTokenizer(br.readLine());

        for(int i=0; i<N; i++)
            arr[i] = Long.parseLong(st.nextToken());

        Arrays.sort(arr);

        for(int i=0; i<N-2; i++){
            int l = i+1;
            int r = N-1;

            while(l < r){
                long sum = arr[i] + arr[l] + arr[r];
                if(mn > Math.abs(sum)){
                    mn = Math.abs(sum);
                    ans[0] = arr[i];
                    ans[1] = arr[l];
                    ans[2] = arr[r];
                }

                if(sum==0) break;
                else if(sum < 0) l++;
                else r--;
            }
        }

        System.out.println(ans[0] + " " + ans[1] + " " + ans[2]);
    }
}
